File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/stack/P739_DailyTemperatures.py
# author: tobebetter9527
# since: 2024/4/29 17:36
from typing import List


class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        n = len(temperatures)
        ans = [0 for _ in range(n)]
        stack = []
        for i in range(n - 1, -1, -1):
            while stack and temperatures[i] >= temperatures[stack[-1]]:
                stack.pop()
            ans[i] = (stack[-1] - i) if stack else 0
            stack.append(i)
        return ans

File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/stack/P739_DailyTemperatures.java
package com.dsa.leetcode.labuladong.stack;

import java.util.ArrayDeque;
import java.util.Deque;

public class P739_DailyTemperatures {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        Deque<Integer> stack = new ArrayDeque<>();
        int[] ans = new int[n];
        for (int i = n - 1; i >= 0; i--) {
            while (!stack.isEmpty() && temperatures[i] >= temperatures[stack.peek()]) {
                stack.pop();
            }
            ans[i] = stack.isEmpty() ? 0 : (stack.peek() - i);
            stack.push(i);
        }
        return ans;
    }

    public static void main(String[] args) {
        P739_DailyTemperatures p739 = new P739_DailyTemperatures();
        int[] temperatures = {73, 74, 75, 71, 69, 72, 76, 73};
        int[] ints = p739.dailyTemperatures(temperatures);
        for (int anInt : ints) {
            System.out.println(anInt);
        }
    }
}
